src/asset_tracking.rs

```diff
@@ -0,0 +1,57 @@
+use bevy::prelude::*;
+use std::collections::VecDeque;
+
+pub(super) fn plugin(app: &mut App) {
+    app.init_resource::<ResourceHandles>()
+        .add_systems(PreUpdate, load_resource_assets);
+}
+
+pub trait LoadResource {
+    /// This will load the [`Resource`] as an [`Asset`]. When all of its asset dependencies
+    /// have been loaded, it will be inserted as a resource. This ensures that the resource only
+    /// exists when the assets are ready.
+    fn load_resource<T: Resource + Asset + Clone + FromWorld>(&mut self) -> &mut Self;
+}
+
+impl LoadResource for App {
+    fn load_resource<T: Resource + Asset + Clone + FromWorld>(&mut self) -> &mut Self {
+        self.init_asset::<T>();
+        let world = self.world_mut();
+        let value = T::from_world(world);
+        let assets = world.resource::<AssetServer>();
+        let handle = assets.add(value);
+        let mut handles = world.resource_mut::<ResourceHandles>();
+        handles
+            .waiting
+            .push_front((handle.untyped(), |world, handle| {
+                let assets = world.resource::<Assets<T>>();
+                if let Some(value) = assets.get(handle.id().typed::<T>()) {
+                    world.insert_resource(value.clone());
+                }
+            }));
+        self
+    }
+}
+
+#[derive(Resource, Default)]
+struct ResourceHandles {
+    waiting: VecDeque<(UntypedHandle, fn(&mut World, &UntypedHandle))>,
```

Comment:
I think `fn(&mut World, &UntypedHandle))` looks pretty scary for a beginner. Either use a simpler type or add a comment making it a bit friendlier.